DOCS-1576: Updates for federated VXLAN and cluster mesh (#940)
tigera/docs: calico-enterprise/multicluster/federation/kubeconfig.mdx

description: Configure a local cluster to pull endpoint data from a remote cluster.
---

# Configure federated endpoint identity and multi-cluster networking

## Big picture

> Create RemoteClusterConfiguration to federate endpoint identities and establish cross-cluster connectivity.

Reviewer: The Big picture should not introduce a specific resource or implementation. User is just trying to see if they are on the right task. How about?

Federate endpoint identities and establish cross-cluster connectivity.

Author: Took your suggestion, but just added that we are "Configuring a cluster..." for ^

### Local and remote clusters

Each cluster in the cluster mesh can act as both a local and remote cluster.

> - Local clusters are configured (via RemoteClusterConfiguration) to retrieve endpoint and routing data from remote clusters

Reviewer: Let's make this parallel with the second bullet:

- Local clusters retrieve endpoint and routing data from remote clusters (via RemoteClusterConfiguration)

- Remote clusters authorize local clusters to retrieve endpoint and routing data

### Remote endpoint identity and policy

> Policy can typically only reference the endpoint identity (e.g. pod labels) of local endpoints. Federated endpoint identity enables local policy rules to reference remote endpoint identities.

Reviewer: ```suggestion
Typically, policy can only reference the endpoint identity (e.g. pod labels) of local endpoints. Federated endpoint identity enables local policy rules to reference remote endpoint identities.
```

### RemoteClusterConfiguration
RemoteClusterConfiguration is the resource that configures a local cluster to sync resources from a remote cluster. It primarily describes how a local cluster establishes that connection to the remote cluster through which resources are synced.

The resources synced through this connection enable the local cluster to reference remote endpoint identity and establish cross-cluster overlay routes.

RemoteClusterConfiguration creates this connection in one direction. If you want identity-aware policy on both sides (i.e. both clusters) of a connection, or you want {{prodname}} to establish cross-cluster overlay networking, you need to create a RemoteClusterConfiguration for both directions.

### kubeconfig files

> Each cluster in the cluster mesh should have a dedicated kubeconfig file used by other clusters in the mesh connect and authenticate.

Reviewer: End of this sentence is a bit unclear "a dedicated kubeconfig file used by other clusters in the mesh connect **and authenticate**"?

Author: I forgot a 'to'!

## Before you begin
**Required**
- [Install {{prodname}}](../../getting-started/install-on-clusters/kubernetes/index.mdx)
- [Ensure pod IP routability](#ensure-pod-ip-routability)

## How to
- [Create kubeconfig files](#create-kubeconfig-files)
- [Create RemoteClusterConfiguration](#create-remoteclusterconfigurations)
- [Validate federation and multi-cluster networking](#validate-federation-and-multi-cluster-networking)
- [Create remote-identity-aware network policy](#create-remote-identity-aware-network-policy)
- [Troubleshoot](#troubleshoot)
- [Configure IP pool resources](#configure-ip-pool-resources)

### Ensure pod IP routability

> Federation of workload endpoint identities only has utility when [Pod IP routability](./overview#pod-ip-routability) is established between clusters. If your clusters are using a supported overlay networking mode, {{prodname}} can automatically meet this requirement when clusters are connected.

Reviewer: ```suggestion
Federation of workload endpoint identities requires [Pod IP routability](./overview#pod-ip-routability) between clusters. If your clusters are using a supported {{prodname}} overlay networking mode, pod IP routability is automatically handled.
```

Author: I've made the suggested change for the first sentence, but I prefer to how the second sentence is currently phrased. Specifically:
* I prefer the less absolute terminology of "can ... meet" over "is .. handled", since the usage of supported overlay mode alone is not enough
* I think noting that this requirement is met "when clusters are connected" helps the user understand what's to be expected in going through this configuration process

Reviewer: I'm fine with that.

   ```bash
   kubectl describe secret -n kube-system $(kubectl get serviceaccounts tigera-federation-remote-cluster -n kube-system -o jsonpath='{.secrets[0].name}')
   ```

1. Retrieve and save the certificate authority and server data:

   Run the following command:
   ```bash
   kubectl config view --flatten --minify
   ```
   Replace `<YOUR-CERTIFICATE-AUTHORITY-DATA>` and `<YOUR-SERVER-ADDRESS>` with `certificate-authority-data` and `server` values respectively.

1. Verify that the `kubeconfig` file works:

   Issue a command like the following to validate the kubeconfig file can be used to connect to the current cluster and access resources:
   ```bash
   kubectl --kubeconfig=kubeconfig-app-a get nodes
   ```

### Create RemoteClusterConfigurations
We'll now create the RemoteClusterConfigurations that establish synchronization between clusters. This enables remote-identity aware policy, federated services, and can establish multi-cluster networking.

**For each pair** of clusters in the cluster mesh (e.g. cluster A and cluster B):

1. In cluster A, create a secret that contains the kubeconfig for cluster B:

>    Determine the namespace you will create the secret in - this will be the value of `<namespace>` to replace in all steps.

Reviewer: ```suggestion
   Determine the namespace (`<namespace>`) for the secret to replace in all steps.
```
